buildscripts/azure/azure-windows.yml

```diff
@@ -43,26 +43,21 @@ jobs:
         call activate %CONDA_ENV%
         conda remove -y tbb tbb-devel
       displayName: 'Remove TBB'
-      condition: eq(variables['PYTHON'], '2.7')
+
+    - script: |
+        # temporarily patch this in to get a recent TBB
+        call activate %CONDA_ENV%
+        conda install -c conda-forge -y tbb tbb-devel
+      displayName: 'Add in conda-forge TBB'
```

Comment:
NOTE TO REVIEWERS: Not to be merged until this is replaced.

Reply:
What needs to be done here?

Reply:
Anaconda distro needs to ship a newer TBB for windows and we need to make it so Numba requires said newer TBB. I believe this is the change needed from TBB: https://github.com/intel/tbb/blob/18070344d755ece04d169e6cc40775cae9288cee/CHANGES#L133-L134 which is in the `2019.5` release but:
```
$ CONDA_SUBDIR=win-64 conda search tbb
Loading channels: done
# Name                       Version           Build  Channel             
tbb                         2018.0.2      h6538335_0  pkgs/main           
tbb                         2018.0.2      hc56fc5f_0  pkgs/main           
tbb                         2018.0.4      hdc96acc_1  pkgs/main           
tbb                         2018.0.4      he980bc4_1  pkgs/main           
tbb                         2018.0.5      hdc96acc_0  pkgs/main           
tbb                         2018.0.5      he980bc4_0  pkgs/main           
tbb                           2019.0      h74a9793_0  pkgs/main           
tbb                           2019.0      hdc96acc_0  pkgs/main           
tbb                           2019.1      h74a9793_0  pkgs/main           
tbb                           2019.1      hdc96acc_0  pkgs/main           
tbb                           2019.4      h74a9793_0  pkgs/main           
tbb                           2019.4      hdc96acc_0  pkgs/main 
```
demonstrates there's no such package for windows.

Reply:
So should we go ahead and update the tbb version requirement in the conda recipe in the build scripts? That's the only place I saw that has a version requirement. It doesn't look like setup.py does a version check. Presumably with a too old tbb the compilation fails.

Reply:
Yes, I think the conditions like these: https://github.com/numba/numba/blob/ef119bcd1733ff49d71bdf2da8a66e91bb704f83/buildscripts/condarecipe.local/meta.yaml#L39
need raising for both `tbb-devel` and `tbb`.

I'm not sure what happens with non-conda things here, a too old `tbb-devel` package at build time would lead to the C++ compiler simply just refusing to compile it. I don't know what would happen at run time if a new `tbb-devel` was used for compilation but an old `tbb` was installed as the run time dependency, I imagine symbol lookup fails or segfaults. I guess `setup.py` needs an `extras` specifying of `tbb` with a constrained version so `pip install numba[tbb]` works as per [example 6.](https://pip.pypa.io/en/latest/reference/pip_install/#examples) and with reference to [these docs](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies) on implementation detail? Is that what you were thinking?

Reply:
I was just thinking of a version check of the library in the part that checks if it can be used. I don't know how easy that is. Anyway setup.py cleanup should probably go in a future PR as that tends to need a lot of testing of its own.

Reply:
Ah I see, I think we'd have to bake the version into a string and expose that, this reasonable to do, if you look at the bottom of the `OpenMP` `ompool.cpp` code you'll see `openmp_vendor` being exposed as a string for a similar purpose (checking if it's GNU OpenMP). `numba.npyufunc.parfor::_launch_threads()` can then check this string if `TBB` is chosen as a backend? How does that sound?

Reply:
I guess you'd have to test it to see if it works. I don't know what happens if you load the wrong tbb library presently.

Reply:
I think you'd probably get a missing symbol like problem.

Reply:
I don't really want to mess with setup.py as I'm not too familiar with all the numba build processes. Should I update the conda recipe or should that not be done until the packages exist on the Anaconda channel?

Reply:
> I think you'd probably get a missing symbol like problem.

I must not understand the suggestion. Would it even be possible to load the library to check the function if the runtime tbb version is wrong? And this doesn't solve the issue of the wrong compile time tbb (the compilation would presumably fail, but it might not be obvious that the version is why without an explicit version check). 

Having the version restriction in the conda metadata would be good as well. It looks like the recipe in this repo is already using run_constrained to do this for an older version, presumably the recipe used to build the Anaconda package does as well. So it just needs to be updated, along with the metadata in the build dependencies. 

As an aside, it looks like some of the metadata in the conda recipe in this repo doesn't exist in the conda-forge recipe, like some of the run_constrained restrictions.

Reply:
> > I think you'd probably get a missing symbol like problem.
> 
> I must not understand the suggestion. Would it even be possible to load the library to check the function if the runtime tbb version is wrong?

I think yes and that this is what needs to be used: https://www.threadingbuildingblocks.org/docs/help/reference/environment/version_information.html but most likely in a separate DSO so that there's no symbols in use other than those to check the interface. Another option would be to add a function to e.g. `_helperlib` that will `dlopen()` with `RTLD_LAZY` a given library, and use that with `ctypes` to bind to the querying functions. I prefer the first way as there's less room for error/reliance on various CPython/Linux internals.

> And this doesn't solve the issue of the wrong compile time tbb (the compilation would presumably fail, but it might not be obvious that the version is why without an explicit version check).

We can check using macros against the TBB version macros and produce a more sensible error message.

> Having the version restriction in the conda metadata would be good as well. It looks like the recipe in this repo is already using run_constrained to do this for an older version, presumably the recipe used to build the Anaconda package does as well. So it just needs to be updated, along with the metadata in the build dependencies.

Yes.

> As an aside, it looks like some of the metadata in the conda recipe in this repo doesn't exist in the conda-forge recipe, like some of the run_constrained restrictions.

This probably needs assessing and propagating upstream, I imagine that conda-forge hasn't e.g. backported fixes to problematic TBBs.

Reply:
Do you want to see these changes to this PR or to future PRs?

Reply:
I worked out an easier way of doing it, see:

* 2889d3d - Add runtime check on TBB version.
* 56a826c - Add TBB version checks at compile time.

Reply:
I ended up doing a few iterations to get everything working, 56a826c through to 46a96f3 on https://github.com/numba/numba/pull/5044. I took the view that this needed to go in as the runtime TBB requirement is really quite recent in relative terms (last 6m).

Reply:
I ended up doing a few iterations to get everything working, 56a826c through to 46a96f3 on #5044. I took the view that this needed to go in as the runtime TBB requirement is really quite recent in relative terms (last 6m).

Reply:
I merged your changes here. I didn't test them all thoroughly but they look ok to me.

Reply:
`tbb` and `tbb-devel` 2020.0 are available for for all platforms in Anaconda's main channel now, this temporary fix can be removed.

Reply:
Fix implemented in https://github.com/numba/numba/pull/5044/commits/baaa14bb0423bc492c04eecf80b80cd793d6e093